webapp/packages/core-resource/src/Resource/CachedResourceOffsetPageKeys.ts

```diff
@@ -108,41 +108,64 @@ export function expandOffsetPageRange(
   outdated: boolean,
   hasNextPage: boolean,
 ): void {
-  const from = info.offset;
-  const to = info.offset + info.limit;
+  const initialFrom = info.offset;
+  const initialTo = info.offset + info.limit;
 
-  let pageInserted = false;
-  for (const page of pages) {
-    if (page.to <= from) {
-      continue;
-    }
+  const newPage = new ResourceOffsetPage().setSize(initialFrom, initialTo).update(info.offset, items).setOutdated(outdated);
 
-    if (!hasNextPage) {
-      if (page.from >= to) {
-        pages.splice(pages.indexOf(page));
-        break;
+  const mergedPages: IResourceOffsetPage[] = [];
+  let i = 0;
+
+  // Add all pages before the newPage
+  while (i < pages.length && pages[i]!.to <= newPage.from) {
+    mergedPages.push(pages[i]!);
+    i++;
+  }
+
+  // Adjust overlapping existing pages
+  while (i < pages.length && pages[i]!.from < newPage.to) {
+    const current = pages[i]!;
+    // If existing page starts before newPage
+    if (current.from < newPage.from) {
+      // Adjust the existing page to end at newPage.from
+      current.setSize(current.from, newPage.from);
+      if (current.to - current.from > 0) {
+        mergedPages.push(current);
       }
     }
-
-    if (page.from <= from && !pageInserted) {
-      if (page.from < from) {
-        page.setSize(page.from, from);
-        pages.splice(pages.indexOf(page) + 1, 0, new ResourceOffsetPage().setSize(from, to).update(from, items).setOutdated(outdated));
-      } else {
-        page.setSize(from, to).update(from, items).setOutdated(outdated);
+    // If existing page ends after newPage
+    if (current.to > newPage.to) {
+      // Adjust the existing page to start at newPage.to
+      current.setSize(newPage.to, current.to);
+      // Since we need to remove pages after newPage when hasNextPage is false,
+      // we only include this adjusted page if hasNextPage is true
+      if (hasNextPage && current.to - current.from > 0) {
+        mergedPages.push(current);
       }
-      pageInserted = true;
-      continue;
     }
+    i++;
+  }
 
-    if (page.isInRange(from, to)) {
-      pages.splice(pages.indexOf(page), 1);
+  // Add the newPage
+  mergedPages.push(newPage);
+
+  // Add the remaining pages after newPage if hasNextPage is true
+  if (hasNextPage) {
+    while (i < pages.length) {
+      mergedPages.push(pages[i]!);
+      i++;
     }
+  } else {
+    // Since hasNextPage is false, we remove all pages after newPage
+    // No action needed here as we simply don't add them
   }
 
-  const lastPage = pages[pages.length - 1];
+  // Remove zero-length ranges
+  const filteredPages = mergedPages.filter(range => range.to - range.from > 0);
```

Comment:
do you think we can simplify the logic upper this line? cause this overlapping logic seems a bit unstable and easy to get a bug here someday

I mean can we do it in 3 cycles:
- Add all pages including `newPage` and do `current.setSize()` (add stage)
- Filter zero-length ranges (filter stage)
- Sort pages by `from` filter (sort stage)

Reply:
I mean that filtering is gonna handle this overlapping logic cause it is ordered